src/lib/Occurrence.hs

```diff
@@ -0,0 +1,623 @@
+  -- we have to assume here that it collides with itself.
+  -- e.g., for i. xs
+  -- Note: If we have a bound on the size of the index set traversed
+  -- by `ib`, then we can compute a bound on the number of accesses,
+  -- but there's no need to do that until we make inlining decisions
+  -- on a more fine-grained basis than "was this used more than once?"
+  if ct == zero then zero else Const Unbounded
+iterate injective ib (At expr use) = At expr' use' where
+  (expr', injective') =
+    flip runState injective $ iterateIxExpr ib expr
+  use' = iterate injective' ib use
+iterate injective ib (Max uses) =
+  foldl max zero $ map (iterate injective ib) uses
+iterate injective ib (Sum uses) =
+  foldl plus zero $ map (iterate injective ib) uses
+
+iterateIxExpr :: NameBinder AtomNameC iter iter'
+              -> IxExpr iter' -> State Bool (IxExpr iter)
+iterateIxExpr ib (Var i) =
+  case hoist ib i of
+    HoistSuccess i' -> return $ Var i'
+    HoistFailure _ ->
+      -- If we hit a reference to the iteration binder,
+      -- that constitutes a proof that indexing is injective,
+      -- because all constructors of `IxExpr` that this could
+      -- be nested under represent invertible functions.
+      -- If we were already injective, this case is an
+      -- over-approximation: the object is actually accessed at just
+      -- one value of the index, but since we don't know which one, we
+      -- assume it was all of them.
+      put True >> return IxAll
+iterateIxExpr ib (Product elts) =
+  -- Rectangular approximation to something like xs.(j, j)
+  Product <$> mapM (iterateIxExpr ib) elts
+iterateIxExpr ib (Inject i elt) =
+  Inject i <$> iterateIxExpr ib elt
+iterateIxExpr ib (Unknown names) =
+  case hoist ib (ListE names) of
+    HoistSuccess (ListE names') -> return $ Unknown names'
+    -- If we hit a reference to an unknown function of the binder, we
+    -- must assume it can hit any data index (which we
+    -- over-approximate as it hitting all of them); and we also do not
+    -- gain a proof that the indexing is injective, because different
+    -- values of the iteration-space binder could lead to the same
+    -- value of the data-space binder.
+    HoistFailure _ -> return IxAll
+iterateIxExpr _ IxAll = return IxAll
+
+-- === CollapsedUse ===
+
+-- CollapsedUse is like Use, but now it is also independent of binders
+-- in scope outside the inlining decision under consideration.
+-- Collapsing Use into CollapsedUse perforce over-approximates those
+-- binders, potentially severely.
+
+-- A CollapsedUse is either a count of the accesses to the whole
+-- object, or a subset of the leading dimension of the object.  In the
+-- latter case, the object is not accessed outside that subset.  The
+-- Subset structure maintains a map of further subsets, and how later
+-- dimensions are accessed in each of them.
+data CollapsedUse =
+    CConst Count
+  | CSubset (Subset CollapsedUse)
```

Comment:
Is `CSubset (SAll (CConst c))` equivalent to `CConst c`?

Reply:
Almost.  They both represent the same amount of access, but `CConst One` corresponds to `xs`, which is work-preserving to inline even if the creation expression is completely opaque; whereas `CSubset (SAll (CConst c))` corresponds to `for i. xs.i`, which is only work-preserving to inline if the outermost dimension is immediately accessible to cancel against the indexing expression.

Reply:
Ah, it might be worth adding this as a comment then. This data structure is not just about the use, but also about the number of indices applied to a value.

Reply:
Comment added.